### PR TITLE
[Refactor] Use local configuration instead of a global one

### PR DESCRIPTION
**Pull request by antstorm**
I'm working on a bigger change that will turn Cadence into an initializable class rather than using a single global instance. This will allow connecting your app to multiple clusters (this is handy for easy migrating from one Cadence cluster to another or more exotic solutions such as client-side sharding).

This is the first step which makes configuration local, passed to all classes that are using it

### CODE REVIEW

File: lib/cadence/execution_options.rb [hunk 1]
```diff
@@ -4,7 +4,7 @@ module Cadence
   class ExecutionOptions
     attr_reader :name, :domain, :task_list, :retry_policy, :timeouts, :headers
 
-    def initialize(object, options = {})
+    def initialize(object, options, defaults = nil)
```
Comment (drewhoskins-stripe) [inline lib/cadence/execution_options.rb:7]: Can you add a comment as to what the shape of `defaults` is and what to pass in?

File: lib/cadence/workflow/decision_task_processor.rb [hunk 2]
```diff
@@ -8,14 +8,15 @@ class Workflow
     class DecisionTaskProcessor
       MAX_FAILED_ATTEMPTS = 50
 
-      def initialize(task, domain, workflow_lookup, client, middleware_chain)
+      def initialize(task, domain, workflow_lookup, connection, middleware_chain, config)
```
Comment (drewhoskins-stripe) [inline lib/cadence/workflow/decision_task_processor.rb:11]: Would love to see some comments in at least some of these classes, like what's a connection, what's a config, or how do I get them to pass in?

File: lib/cadence/workflow/poller.rb [hunk 3]
```diff
@@ -68,17 +69,19 @@ def poll_loop
...
-        client.poll_for_decision_task(domain: domain, task_list: task_list)
+        connection.poll_for_decision_task(domain: domain, task_list: task_list)
       rescue StandardError => error
         Cadence.logger.error("Unable to poll for a decision task: #{error.inspect}")
         nil
       end
 
       def process(task)
-        client = Cadence::Client.generate
+        connection = Cadence::Connection.generate(config.for_connection)
```
Comment (drewhoskins-stripe) [inline lib/cadence/workflow/poller.rb:79]: could/should this be done inside of `DecisionTaskProcessor.initialize`?

**Review by drewhoskins-stripe — COMMENTED** [review]
Bullish on the direction!
Looks good as far as I can tell, though could use some documentation.

File: lib/cadence/configuration.rb [hunk 4]
```diff
@@ -32,5 +35,22 @@ def initialize
...
+
+    def for_connection
+      Connection.new(
+        type: connection_type,
+        host: host,
+        port: port
+      ).freeze
+    end
+
+    def for_execution
```
Comment (jeffschoner-stripe) [inline lib/cadence/configuration.rb:47]: This seems to only be used for passing in defaults to `ExecutionOptions.new`. Should this be named `default_execution_options` instead?

File: lib/cadence/execution_options.rb (hunk 1 above)
Comment (antstorm) [reply lib/cadence/execution_options.rb:7]: It is an instance of `Cadence::Configuration::Execution`, but I see your point, this is immediately visible and can cause confusion. Agree that we need to address it

File: lib/cadence/workflow/decision_task_processor.rb (hunk 2 above)
Comment (antstorm) [reply lib/cadence/workflow/decision_task_processor.rb:11]: Yes, point taken, will think of a solution 👍

File: lib/cadence/workflow/poller.rb (hunk 3 above)
Comment (antstorm) [reply lib/cadence/workflow/poller.rb:79]: We can do this, yes. The benefit of this approach is dependency injection, but I think this one makes sense to move into the `DecisionTaskProcessor`. I'll fix it, thanks

File: lib/cadence/connection.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    CLIENT_TYPES_MAP = {
+      thrift: Cadence::Connection::Thrift
+    }.freeze
+
+    def self.generate(configuration, options = {})
+      connection_class = CLIENT_TYPES_MAP[configuration.type]
+      host = configuration.host
+      port = configuration.port
+
+      hostname = `hostname`
```
[the remark this answers is not in the document]
Comment (antstorm) [reply lib/cadence/connection.rb:14]: Oh, nice, didn't know of this one. I'll add it as a follow-up PR (we had `\`hostname\`` for a while and I want to keep changes related to config in this PR)

**Review by jeffschoner-stripe — APPROVED** [review]

**Review by cedric-cordenier — APPROVED** [review]

**Pull request merged**